README.md
# space-invaders

[![Build Status](https://github.com/dkim/space-invaders/workflows/build/badge.svg)](https://github.com/dkim/space-invaders/actions?query=workflow%3Abuild+branch%3Amaster)

space-invaders is an OpenGL-based emulator in Rust for the [Space Invaders]
arcade machine.

[Space Invaders]: https://en.wikipedia.org/wiki/Space_Invaders

![A Screenshot of space-invaders](screenshot.png "A Screenshot of space-invaders")

## Installation Requirements

This program targets the latest stable version of Rust 1.45.2 or later.

### Requirements for Linux Systems

#### GLFW's Dependencies

This program uses the [`glfw-sys`] package (via the [`luminance-glfw`]
package), which compiles its own copy of the [GLFW] library. Compiling GLFW on
Linux requires [CMake] and some X11 development libraries to be installed. You
can install them on Debian-based distributions, as follows:

```console
$ sudo apt install cmake libxcursor-dev libxi-dev libxinerama-dev libxrandr-dev
```

The "[Compiling GLFW]" guide provides full details on how to compile GLFW.

[`glfw-sys`]: https://crates.io/crates/glfw-sys
[`luminance-glfw`]: https://crates.io/crates/luminance-glfw
[GLFW]: https://www.glfw.org
[CMake]: https://cmake.org
[Compiling GLFW]: https://www.glfw.org/docs/latest/compile.html

#### CPAL's Dependencies

This program uses [CPAL], a cross-platform audio library, via the [`rodio`]
package. On Linux, CPAL depends on the [`alsa-sys`] package, which requires
[pkg-config] and the ALSA development files to be installed. You can install
them on Debian-based distributions, as follows:

```console
$ sudo apt install pkg-config libasound2-dev
```

[CPAL's `README.md`] provides more information on its external library
dependencies.

[CPAL]: https://crates.io/crates/cpal
[`rodio`]: https://crates.io/crates/rodio
[`alsa-sys`]: https://crates.io/crates/alsa-sys
[pkg-config]: https://www.freedesktop.org/wiki/Software/pkg-config/
[CPAL's `README.md`]: https://github.com/RustAudio/cpal/blob/master/README.md

### Requirements for macOS Systems

Xcode contains all the necessary header files, libraries, and tools except for
[CMake]. You can install CMake via [Homebrew], as follows:

```console
$ brew install cmake
```

[Homebrew]: https://brew.sh

### Requirements for Windows Systems

Visual C++ with the "[C++ CMake tools for Windows]" and "[Windows 10 SDK]"
components contains all the necessary header files, libraries, and tools.

[C++ CMake tools for Windows]: https://docs.microsoft.com/en-us/cpp/build/cmake-projects-in-visual-studio?view=vs-2019
[Windows 10 SDK]: https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk

## Installation

```console
$ git clone https://github.com/dkim/space-invaders.git
$ cd space-invaders
$ cargo update  # optional
$ cargo build --release
```

## Usage

```console
$ cargo run --release -- --help
space-invaders 1.1.0
space-invaders is an emulator for the Space Invaders arcade machine.

USAGE:
    space-invaders <roms> [samples]

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

ARGS:
    <roms>       A directory that contains invaders.{e,f,g,h}
    <samples>    A directory that contains {0..8}.wav

$ RUST_LOG=info cargo run --release -- /path/to/roms /path/to/samples
```

### Controls

| Key              | Description                                        |
| ---------------- | -------------------------------------------------- |
| <kbd>←</kbd>     | Move Left                                          |
| <kbd>→</kbd>     | Move Right                                         |
| <kbd>Space</kbd> | Fire                                               |
| <kbd>C</kbd>     | Insert a coin                                      |
| <kbd>T</kbd>     | Tilt the machine                                   |
| <kbd>1</kbd>     | Start a game in single-player mode                 |
| <kbd>2</kbd>     | Start a game in two-player mode                    |
| <kbd>F1</kbd>    | Number of lives: 3 (default) / 4 / 5 / 6           |
| <kbd>F2</kbd>    | Extra life at: 1000 points / 1500 points (default) |
| <kbd>F3</kbd>    | Pricing display: on (default) / off                |

## License

Licensed under either of

 * Apache License, Version 2.0
   ([LICENSE-APACHE](LICENSE-APACHE) or https://www.apache.org/licenses/LICENSE-2.0)
 * MIT license
   ([LICENSE-MIT](LICENSE-MIT) or https://opensource.org/licenses/MIT)

at your option.

## Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed

- Construct a mapping table from bytes to texels at compile time.

## [1.1.0] - 2020-08-15

### Changed

- Simplify error handling in `main.rs` with `Box<dyn std::error::Error>`.
- Upgrade `luminance` to version 0.41.0.

## [1.0.0] - 2020-07-26

[Unreleased]: https://github.com/dkim/space-invaders/compare/1.1.0...HEAD
[1.1.0]: https://github.com/dkim/space-invaders/compare/1.0.0...1.1.0
[1.0.0]: https://github.com/dkim/space-invaders/releases/tag/1.0.0


src/lib.rs
#![warn(rust_2018_idioms)]

use std::{
    fmt::{self, Display, Formatter},
    fs,
    io::{self, Cursor},
    path::Path,
    sync::mpsc::{Receiver, TryRecvError},
};

use bitflags::bitflags;

use log::warn;

use rodio::{Decoder, Device, Sink, Source};

use i8080::Intel8080;

/// An error that can occur in this crate.
#[derive(Debug)]
pub enum Error {
    /// An error from crate `i8080`.
    I8080 { source: i8080::Error },
    /// An I/O error.
    Io { source: io::Error },
}

impl Display for Error {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            Error::I8080 { source } => source.fmt(f),
            Error::Io { source } => source.fmt(f),
        }
    }
}

impl std::error::Error for Error {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            Error::I8080 { source } => Some(source),
            Error::Io { source } => Some(source),
        }
    }
}

impl From<i8080::Error> for Error {
    fn from(e: i8080::Error) -> Self {
        Error::I8080 { source: e }
    }
}

impl From<io::Error> for Error {
    fn from(e: io::Error) -> Self {
        Error::Io { source: e }
    }
}

/// A specialized `std::result::Result` type for this crate.
pub type Result<T> = std::result::Result<T, Error>;

/// The width of the screen of the Space Invaders arcade machine.
pub const SCREEN_WIDTH: u32 = 224;
/// The height of the screen of the Space Invaders arcade machine.
pub const SCREEN_HEIGHT: u32 = 256;

/// A Space Invaders arcade machine.
pub struct SpaceInvaders {
    /// The Intel 8080 CPU.
    pub i8080: Intel8080,
    interrupt_receiver: Receiver<[u8; 3]>,
    /// Port 1.
    pub port1: Port1,
    /// Port 2.
    pub port2: Port2,
    port3: Port3,
    port5: Port5,
    video_shifter: VideoShifter,
    audio_device: Option<Device>,
    samples: Samples,
}

impl SpaceInvaders {
    /// Constructs a new `SpaceInvaders`.
    ///
    /// # Arguments
    ///
    /// * `roms` - a reference to a slice of paths to ROMs to be loaded starting at address 0.
    /// * `samples` - an optional array of paths to 9 audio samples.
    /// * `interrupt_receiver` - a `std::sync::mpsc::Receiver` to receive interrupts from.
    ///
    /// # Example
    ///
    /// ```no_run
    /// use std::sync::mpsc;
    /// use space_invaders::SpaceInvaders;
    ///
    /// let (interrupt_sender, interrupt_receiver) = mpsc::sync_channel(0);
    /// let space_invaders = SpaceInvaders::new(
    ///     &["invaders.h", "invaders.g", "invaders.f", "invaders.e"],
    ///     Some([
    ///         "1.wav", "2.wav", "3.wav", "4.wav", "5.wav", "6.wav", "7.wav", "8.wav", "9.wav",
    ///     ]),
    ///     interrupt_receiver,
    /// )?;
    /// # Ok::<(), space_invaders::Error>(())
    /// ```
    pub fn new<P: AsRef<Path>, Q: AsRef<Path>>(
        roms: &[P],
        samples: Option<[Q; 9]>,
        interrupt_receiver: Receiver<[u8; 3]>,
    ) -> Result<Self> {
        let audio_device = rodio::default_output_device();
        let samples = Samples::new(&audio_device, samples);
        Ok(Self {
            i8080: Intel8080::new(roms, 0)?,
            interrupt_receiver,
            port1: Port1::default(),
            port2: Port2::default(),
            port3: Port3::default(),
            port5: Port5::default(),
            video_shifter: VideoShifter::default(),
            audio_device,
            samples,
        })
    }

    /// Returns a shared reference to the framebuffer.
    pub fn framebuffer(&self) -> &[u8] {
        &self.i8080.memory[0x2400..0x4000]
    }

    /// Handles a pending interrupt, if any; otherwise fetches and executes an instruction.
    pub fn update(&mut self) -> u32 {
        match self.interrupt_receiver.try_recv() {
            Err(TryRecvError::Empty) | Err(TryRecvError::Disconnected) => {
                self.fetch_execute_instruction()
            }
            Ok(instruction) => self.i8080.interrupt(instruction).unwrap_or(0),
        }
    }

    fn fetch_execute_instruction(&mut self) -> u32 {
        let (instruction, states) = self.i8080.fetch_execute_instruction();
        match instruction {
            // OUT port
            [0xD3, port, 0] => match port {
                2 => self.video_shifter.offset = self.i8080.cpu.a,
                3 => {
                    // from_bits_unchecked() is used instead of from_bits() because the
                    // functionalities of some bits of port 3 are not clear and they are ignored
                    // for now.
                    let port3 = unsafe { Port3::from_bits_unchecked(self.i8080.cpu.a) };
                    if let Some((ref wav, ref mut sink)) = self.samples.ufo_low_pitch {
                        if port3.contains(Port3::UFO_LOW_PITCH) {
                            if !self.port3.contains(Port3::UFO_LOW_PITCH) {
                                match Decoder::new(Cursor::new(wav.clone())) {
                                    Ok(source) => sink.append(source.repeat_infinite()),
                                    Err(err) => warn!("{:?}", err),
                                }
                            }
                        } else if self.port3.contains(Port3::UFO_LOW_PITCH) {
                            // Sink::stop() cannot be used here. Sink cannot be restarted once it
                            // has been stopped. Refer to
                            // https://github.com/RustAudio/rodio/issues/171.
                            *sink = Sink::new(self.audio_device.as_ref().unwrap());
                        }
                    }
                    for (audio, bit) in &mut [
                        (&self.samples.shoot, Port3::SHOOT),
                        (&self.samples.explosion, Port3::EXPLOSION),
                        (&self.samples.invader_killed, Port3::INVADER_KILLED),
                    ] {
                        if let Some((wav, sink)) = audio {
                            if port3.contains(*bit) && !self.port3.contains(*bit) {
                                match Decoder::new(Cursor::new(wav.clone())) {
                                    Ok(source) => sink.append(source),
                                    Err(err) => warn!("{:?}", err),
                                }
                            }
                        }
                    }
                    self.port3 = port3;
                }
                4 => self.video_shifter.shift_right(self.i8080.cpu.a),
                5 => {
                    let port5 = Port5::from_bits(self.i8080.cpu.a).unwrap();
                    for (audio, bit) in &mut [
                        (&self.samples.fast_invader_1, Port5::FAST_INVADER_1),
                        (&self.samples.fast_invader_2, Port5::FAST_INVADER_2),
                        (&self.samples.fast_invader_3, Port5::FAST_INVADER_3),
                        (&self.samples.fast_invader_4, Port5::FAST_INVADER_4),
                        (&self.samples.ufo_high_pitch, Port5::UFO_HIGH_PITCH),
                    ] {
                        if let Some((wav, sink)) = audio {
                            if port5.contains(*bit) && !self.port5.contains(*bit) {
                                match Decoder::new(Cursor::new(wav.clone())) {
                                    Ok(source) => sink.append(source),
                                    Err(err) => warn!("{:?}", err),
                                }
                            }
                        }
                    }
                    self.port5 = port5;
                }
                6 => (), // watchdog
                _ => unreachable!(),
            },
            // IN port
            [0xDB, port, 0] => match port {
                1 => self.i8080.cpu.a = self.port1.bits(),
                2 => self.i8080.cpu.a = self.port2.bits(),
                3 => self.i8080.cpu.a = self.video_shifter.into(),
                _ => unreachable!(),
            },
            _ => (),
        }
        states
    }
}

bitflags! {
    /// Port 1, which consists of bit flags.
    pub struct Port1: u8 {
        const COIN = 0b0000_0001;
        const PLAYER_2_START = 0b0000_0010;
        const PLAYER_1_START = 0b0000_0100;
        const ALWAYS_ONE = 0b0000_1000;
        const PLAYER_1_FIRE = 0b0001_0000;
        const PLAYER_1_LEFT = 0b0010_0000;
        const PLAYER_1_RIGHT = 0b0100_0000;
    }
}

impl Default for Port1 {
    fn default() -> Self {
        Port1::ALWAYS_ONE
    }
}

bitflags! {
    /// Port 2, which consists of bit flags.
    #[derive(Default)]
    pub struct Port2: u8 {
        const TILT = 0b0000_0100;
        const EXTRA_LIFE_AT = 0b0000_1000;
        const PLAYER_2_FIRE = 0b0001_0000;
        const PLAYER_2_LEFT = 0b0010_0000;
        const PLAYER_2_RIGHT = 0b0100_0000;
        const PRICING_DISPLAY = 0b1000_0000;
    }
}

bitflags! {
    // Some bits of port 3 are missing here because their functionalities are not clear.
    #[derive(Default)]
    struct Port3: u8 {
        const UFO_LOW_PITCH = 0b0000_0001;
        const SHOOT = 0b0000_0010;
        const EXPLOSION = 0b0000_0100;
        const INVADER_KILLED = 0b0000_1000;
    }
}

bitflags! {
    #[derive(Default)]
    struct Port5: u8 {
        const FAST_INVADER_1 = 0b0000_0001;
        const FAST_INVADER_2 = 0b0000_0010;
        const FAST_INVADER_3 = 0b0000_0100;
        const FAST_INVADER_4 = 0b0000_1000;
        const UFO_HIGH_PITCH = 0b0001_0000;
    }
}

#[derive(Clone, Copy, Default)]
struct VideoShifter {
    register: u16,
    offset: u8,
}

impl VideoShifter {
    fn shift_right(&mut self, fill_byte: u8) {
        self.register = (u16::from(fill_byte) << 8) | (self.register >> 8);
    }
}

impl Into<u8> for VideoShifter {
    fn into(self) -> u8 {
        (self.register >> (8 - self.offset)) as u8
    }
}

struct Samples {
    ufo_low_pitch: Option<(Vec<u8>, Sink)>,
    shoot: Option<(Vec<u8>, Sink)>,
    explosion: Option<(Vec<u8>, Sink)>,
    invader_killed: Option<(Vec<u8>, Sink)>,
    fast_invader_1: Option<(Vec<u8>, Sink)>,
    fast_invader_2: Option<(Vec<u8>, Sink)>,
    fast_invader_3: Option<(Vec<u8>, Sink)>,
    fast_invader_4: Option<(Vec<u8>, Sink)>,
    ufo_high_pitch: Option<(Vec<u8>, Sink)>,
}

impl Samples {
    fn new<P: AsRef<Path>>(audio_device: &Option<Device>, samples: Option<[P; 9]>) -> Self {
        let mut ufo_low_pitch = None;
        let mut shoot = None;
        let mut explosion = None;
        let mut invader_killed = None;
        let mut fast_invader_1 = None;
        let mut fast_invader_2 = None;
        let mut fast_invader_3 = None;
        let mut fast_invader_4 = None;
        let mut ufo_high_pitch = None;
        if let Some(samples) = samples {
            if let Some(audio_device) = audio_device {
                for (path, audio) in &mut [
                    (&samples[0], &mut ufo_high_pitch),
                    (&samples[1], &mut shoot),
                    (&samples[2], &mut explosion),
                    (&samples[3], &mut invader_killed),
                    (&samples[4], &mut fast_invader_1),
                    (&samples[5], &mut fast_invader_2),
                    (&samples[6], &mut fast_invader_3),
                    (&samples[7], &mut fast_invader_4),
                    (&samples[8], &mut ufo_low_pitch),
                ] {
                    let path = path.as_ref();
                    match fs::read(path) {
                        Ok(wav) => **audio = Some((wav, Sink::new(audio_device))),
                        Err(err) => {
                            if let io::ErrorKind::NotFound = err.kind() {
                                warn!("{:?}: '{}'", err, path.display());
                            } else {
                                warn!("{:?}", err);
                            }
                        }
                    }
                }
            }
        }
        Self {
            ufo_low_pitch,
            shoot,
            explosion,
            invader_killed,
            fast_invader_1,
            fast_invader_2,
            fast_invader_3,
            fast_invader_4,
            ufo_high_pitch,
        }
    }
}


src/main.rs
#![warn(rust_2018_idioms)]

use std::{
    mem::MaybeUninit,
    path::PathBuf,
    process,
    sync::{
        mpsc::{self, SyncSender},
        Arc, Mutex,
    },
    thread,
};

use log::info;

use glfw::{Action, Context, Key, WindowEvent};
use luminance_derive::UniformInterface;
use luminance_front::{
    context::GraphicsContext,
    framebuffer::{Framebuffer, FramebufferError},
    pipeline::{PipelineState, TextureBinding},
    pixel::{NormR8UI, NormUnsigned, Pixel},
    render_state::RenderState,
    shader::{BuiltProgram, Program, Uniform},
    tess::{Mode, Tess, TessBuilder},
    texture::{Dim2, GenMipmaps, Sampler, Texture},
};
use luminance_glfw::GlfwSurface;
use luminance_windowing::{WindowDim, WindowOpt};

use spin_sleep::LoopHelper;

use structopt::StructOpt;

use space_invaders::{Port1, Port2, SpaceInvaders};

#[derive(Debug, StructOpt)]
#[structopt(about)]
struct Opt {
    /// A directory that contains invaders.{e,f,g,h}
    #[structopt(parse(from_os_str))]
    roms: PathBuf,

    /// A directory that contains {0..8}.wav
    #[structopt(parse(from_os_str))]
    samples: Option<PathBuf>,
}

#[derive(UniformInterface)]
struct Uniforms {
    sampler: Uniform<TextureBinding<Dim2, NormUnsigned>>,
}

const VERTEX_SHADER: &str = include_str!("vertex.vert");
const FRAGMENT_SHADER: &str = include_str!("fragment.frag");

const FRAMEBUFFER_LEN: usize =
    space_invaders::SCREEN_HEIGHT as usize / 8 * space_invaders::SCREEN_WIDTH as usize;
const TEXELS_LEN: usize =
    space_invaders::SCREEN_HEIGHT as usize * space_invaders::SCREEN_WIDTH as usize;

fn main() {
    if let Err(err) = run(Opt::from_args()) {
        eprintln!("Error: {}", err);
        process::exit(1);
    }
}

fn run(opt: Opt) -> Result<(), Box<dyn std::error::Error>> {
    env_logger::init();

    let (interrupt_sender, interrupt_receiver) = mpsc::sync_channel(0);
    let space_invaders = Arc::new(Mutex::new(SpaceInvaders::new(
        &[
            opt.roms.join("invaders.h"),
            opt.roms.join("invaders.g"),
            opt.roms.join("invaders.f"),
            opt.roms.join("invaders.e"),
        ],
        opt.samples.map(|samples| {
            [
                samples.join("0.wav"),
                samples.join("1.wav"),
                samples.join("2.wav"),
                samples.join("3.wav"),
                samples.join("4.wav"),
                samples.join("5.wav"),
                samples.join("6.wav"),
                samples.join("7.wav"),
                samples.join("8.wav"),
            ]
        }),
        interrupt_receiver,
    )?));
    thread::spawn(update_space_invaders(Arc::clone(&space_invaders)));
    thread::spawn(generate_interrupts(interrupt_sender));

    let mut surface = GlfwSurface::new_gl33(
        "Space Invaders",
        WindowOpt::default().set_dim(WindowDim::Windowed {
            width: space_invaders::SCREEN_WIDTH * 2,
            height: space_invaders::SCREEN_HEIGHT * 2,
        }),
    )?;
    let mut graphics = Graphics::new(&mut surface)?;

    let mut loop_helper = LoopHelper::builder().build_with_target_rate(60.0);
    loop {
        loop_helper.loop_start();
        if !(process_input(&mut surface, &mut graphics, &space_invaders)?) {
            break;
        }
        graphics.render(&space_invaders, &mut surface)?;
        loop_helper.loop_sleep();
    }
    Ok(())
}

fn update_space_invaders(space_invaders: Arc<Mutex<SpaceInvaders>>) -> impl FnOnce() {
    move || {
        let mut loop_helper = LoopHelper::builder().build_with_target_rate(120.0);
        loop {
            // 2 MHz = 2,000,000 states per second = 2 states per microsecond
            let elapsed_states = loop_helper.loop_start().as_micros() * 2;
            let mut states = 0;
            while elapsed_states > states {
                states += u128::from(space_invaders.lock().unwrap().update());
            }
            loop_helper.loop_sleep();
        }
    }
}

fn generate_interrupts(interrupt_sender: SyncSender<[u8; 3]>) -> impl FnOnce() {
    move || {
        let mut loop_helper = LoopHelper::builder().build_with_target_rate(120.0);
        loop {
            loop_helper.loop_start();
            if interrupt_sender.send([0xCF, 0, 0] /* RST 1 */).is_err() {
                break;
            }
            loop_helper.loop_sleep();
            loop_helper.loop_start();
            if interrupt_sender.send([0xD7, 0, 0] /* RST 2 */).is_err() {
                break;
            }
            loop_helper.loop_sleep();
        }
    }
}

struct Graphics {
    back_buffer: Framebuffer<Dim2, (), ()>,
    pipeline_state: PipelineState,
    program: Program<(), (), Uniforms>,
    render_state: RenderState,
    vertices: Tess<()>,
    texture: Texture<Dim2, NormR8UI>,
    texels: [<NormR8UI as Pixel>::Encoding; TEXELS_LEN],
}

impl Graphics {
    fn new(surface: &mut GlfwSurface) -> Result<Self, Box<dyn std::error::Error>> {
        let back_buffer = surface.back_buffer()?;
        let pipeline_state = PipelineState::default().enable_clear_depth(false);
        let BuiltProgram { program, warnings } =
            surface.new_shader_program::<(), (), Uniforms>().from_strings(
                VERTEX_SHADER,
                None, // tessellation shaders
                None, // geometry shader
                FRAGMENT_SHADER,
            )?;
        assert!(warnings.is_empty(), "{:?}", warnings);
        let render_state = RenderState::default().set_depth_test(None);
        let vertices =
            TessBuilder::new(surface).set_vertex_nb(4).set_mode(Mode::TriangleFan).build()?;
        let texture = Texture::<Dim2, NormR8UI>::new(
            surface,
            [space_invaders::SCREEN_HEIGHT, space_invaders::SCREEN_WIDTH],
            0, // mipmaps
            Sampler::default(),
        )?;
        let texels = [0; TEXELS_LEN];
        Ok(Self { back_buffer, pipeline_state, program, render_state, vertices, texture, texels })
    }

    fn render(
        &mut self,
        space_invaders: &Mutex<SpaceInvaders>,
        surface: &mut GlfwSurface,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let Graphics {
            back_buffer,
            pipeline_state,
            program,
            render_state,
            vertices,
            texture,
            texels,
        } = self;

        let framebuffer = unsafe {
            let mut framebuffer = MaybeUninit::<[u8; FRAMEBUFFER_LEN]>::uninit();
            (framebuffer.as_mut_ptr() as *mut u8).copy_from_nonoverlapping(
                space_invaders.lock().unwrap().framebuffer() as *const [u8] as *const u8,
                FRAMEBUFFER_LEN,
            );
            framebuffer.assume_init()
        };
        framebuffer_to_texels(&framebuffer, texels);
        texture.upload_raw(GenMipmaps::No, texels)?;
        surface
            .new_pipeline_gate()
            .pipeline(back_buffer, pipeline_state, |pipeline, mut shading_gate| {
                let bound_texture = pipeline.bind_texture(texture)?;
                shading_gate.shade(program, |mut program_interface, uniforms, mut render_gate| {
                    program_interface.set(&uniforms.sampler, bound_texture.binding());
                    render_gate.render(render_state, |mut tess_gate| tess_gate.render(&*vertices))
                })
            })
            .assume()
            .into_result()?;
        surface.window.swap_buffers();
        Ok(())
    }
}

const BYTE_TO_TEXELS: [[u8; 8]; 256] = byte_to_texels();

const fn byte_to_texels() -> [[u8; 8]; 256] {
    let mut mapping = [[0; 8]; 256];
    let mut i = 0;
    while i < 256 {
        let byte = i as u8;
        mapping[i] = [
            if byte & 0x01 > 0 { 0xFF } else { 0x00 },
            if byte & 0x02 > 0 { 0xFF } else { 0x00 },
            if byte & 0x04 > 0 { 0xFF } else { 0x00 },
            if byte & 0x08 > 0 { 0xFF } else { 0x00 },
            if byte & 0x10 > 0 { 0xFF } else { 0x00 },
            if byte & 0x20 > 0 { 0xFF } else { 0x00 },
            if byte & 0x40 > 0 { 0xFF } else { 0x00 },
            if byte & 0x80 > 0 { 0xFF } else { 0x00 },
        ];
        i += 1;
    }
    mapping
}

fn framebuffer_to_texels(
    framebuffer: &[u8],
    texels: &mut [<NormR8UI as Pixel>::Encoding; TEXELS_LEN],
) {
    framebuffer.iter().enumerate().for_each(|(i, &byte)| {
        texels[i * 8..(i + 1) * 8].copy_from_slice(&BYTE_TO_TEXELS[usize::from(byte)]);
    });
}

fn process_input(
    surface: &mut GlfwSurface,
    graphics: &mut Graphics,
    space_invaders: &Mutex<SpaceInvaders>,
) -> Result<bool, FramebufferError> {
    let mut resized = false;
    surface.window.glfw.poll_events();
    for (_, event) in surface.events_rx.try_iter() {
        match event {
            WindowEvent::Key(Key::Left, _, action, _) => match action {
                Action::Press => {
                    let mut space_invaders = space_invaders.lock().unwrap();
                    space_invaders.port1.insert(Port1::PLAYER_1_LEFT);
                    space_invaders.port2.insert(Port2::PLAYER_2_LEFT);
                }
                Action::Release => {
                    let mut space_invaders = space_invaders.lock().unwrap();
                    space_invaders.port1.remove(Port1::PLAYER_1_LEFT);
                    space_invaders.port2.remove(Port2::PLAYER_2_LEFT);
                }
                Action::Repeat => (),
            },
            WindowEvent::Key(Key::Right, _, action, _) => match action {
                Action::Press => {
                    let mut space_invaders = space_invaders.lock().unwrap();
                    space_invaders.port1.insert(Port1::PLAYER_1_RIGHT);
                    space_invaders.port2.insert(Port2::PLAYER_2_RIGHT);
                }
                Action::Release => {
                    let mut space_invaders = space_invaders.lock().unwrap();
                    space_invaders.port1.remove(Port1::PLAYER_1_RIGHT);
                    space_invaders.port2.remove(Port2::PLAYER_2_RIGHT);
                }
                Action::Repeat => (),
            },
            WindowEvent::Key(Key::Space, _, action, _) => match action {
                Action::Press => {
                    let mut space_invaders = space_invaders.lock().unwrap();
                    space_invaders.port1.insert(Port1::PLAYER_1_FIRE);
                    space_invaders.port2.insert(Port2::PLAYER_2_FIRE);
                }
                Action::Release => {
                    let mut space_invaders = space_invaders.lock().unwrap();
                    space_invaders.port1.remove(Port1::PLAYER_1_FIRE);
                    space_invaders.port2.remove(Port2::PLAYER_2_FIRE);
                }
                Action::Repeat => (),
            },
            WindowEvent::Key(Key::C, _, action, _) => match action {
                Action::Press => {
                    space_invaders.lock().unwrap().port1.insert(Port1::COIN);
                }
                Action::Release => {
                    space_invaders.lock().unwrap().port1.remove(Port1::COIN);
                }
                Action::Repeat => (),
            },
            WindowEvent::Key(Key::T, _, action, _) => match action {
                Action::Press => {
                    space_invaders.lock().unwrap().port2.insert(Port2::TILT);
                }
                Action::Release => {
                    space_invaders.lock().unwrap().port2.remove(Port2::TILT);
                }
                Action::Repeat => (),
            },
            WindowEvent::Key(Key::Num1, _, action, _) => match action {
                Action::Press => {
                    space_invaders.lock().unwrap().port1.insert(Port1::PLAYER_1_START);
                }
                Action::Release => {
                    space_invaders.lock().unwrap().port1.remove(Port1::PLAYER_1_START);
                }
                Action::Repeat => (),
            },
            WindowEvent::Key(Key::Num2, _, action, _) => match action {
                Action::Press => {
                    space_invaders.lock().unwrap().port1.insert(Port1::PLAYER_2_START);
                }
                Action::Release => {
                    space_invaders.lock().unwrap().port1.remove(Port1::PLAYER_2_START);
                }
                Action::Repeat => (),
            },
            WindowEvent::Key(Key::F1, _, action, _) => match action {
                Action::Press => {
                    let mut space_invaders = space_invaders.lock().unwrap();
                    let mut bits = space_invaders.port2.bits();
                    bits = (bits & 0b1111_1100) | (((bits & 0b0000_0011) + 1) % 4);
                    space_invaders.port2 = unsafe { Port2::from_bits_unchecked(bits) };
                    match space_invaders.port2.bits() & 0b0000_0011 {
                        0 => info!("num of lives: 3"),
                        1 => info!("num of lives: 4"),
                        2 => info!("num of lives: 5"),
                        3 => info!("num of lives: 6"),
                        _ => unreachable!(),
                    }
                }
                Action::Release | Action::Repeat => (),
            },
            WindowEvent::Key(Key::F2, _, action, _) => match action {
                Action::Press => {
                    let mut space_invaders = space_invaders.lock().unwrap();
                    space_invaders.port2.toggle(Port2::EXTRA_LIFE_AT);
                    if space_invaders.port2.contains(Port2::EXTRA_LIFE_AT) {
                        info!("extra life at: 1000 points");
                    } else {
                        info!("extra life at: 1500 points");
                    }
                }
                Action::Release | Action::Repeat => (),
            },
            WindowEvent::Key(Key::F3, _, action, _) => match action {
                Action::Press => {
                    let mut space_invaders = space_invaders.lock().unwrap();
                    space_invaders.port2.toggle(Port2::PRICING_DISPLAY);
                    if space_invaders.port2.contains(Port2::PRICING_DISPLAY) {
                        info!("pricing display: off");
                    } else {
                        info!("pricing display: on");
                    }
                }
                Action::Release | Action::Repeat => (),
            },
            WindowEvent::FramebufferSize(_, _) => resized = true,
            WindowEvent::Close => return Ok(false),
            _ => (),
        }
    }
    if resized {
        graphics.back_buffer = surface.back_buffer()?;
    }
    Ok(true)
}
